[routing] [strings] Add the street name to verbal turn instructions (#3130)
organicmaps/organicmaps: data/strings/sound.txt

    mr = उजवीकडे वळा
    nl = Sla rechtsaf
    pl = Skręć w prawo
    pt = Vire à direita
    pt-BR = Vire à direita
    ro = Virați la dreapta
    ru = Поверните направо
    sk = Odbočte doprava
    sv = Sväng höger
    sw = Zunguka upande wa kulia
    th = เลี้ยวขวา
    tr = Sağa dönün
    uk = Поверніть праворуч
    vi = Rẽ phải
    zh-Hans = 右转
    zh-Hant = 右轉

  [make_a_sharp_right_turn]
    en = Sharp right
    ar = إلى أقصى اليمين
    be = Рэзка павярніце направа
    cs = Ostře doprava
    da = Skarpt til højre
    de = Scharf rechts
    el = Απότομα δεξιά
    es = Gire bruscamente a la derecha
    eu = Eman bira itxia eskuinera
    fa = منتها الیه سمت راست
    fi = Jyrkkä oikea
    fr = À droite toute
    hi = तीव्रता से दाएं मुड़े
    hr = Oštro udesno
    hu = Kanyarodjon élesen jobbra
    id = Belok tajam ke kanan
    it = Svoltare a destra
    ja = 急に右折
    ko = 급우회전
    mr = तीव्र उजवे वळण
    nl = Scherp rechtsaf
    pl = Ostro w prawo
    pt = Curva acentuada à direita
    pt-BR = Curva acentuada à direita
    ro = Virați strâns la dreapta
    ru = Резко поверните направо
    sk = Ostro doprava
    sv = Skarpt höger
    sw = Fanya kulia mkali
    th = หักมุมขวา

>     tr = Tam sağa dönün

Reviewer: It should be easy to programmatically remove the last dot (all unicode dots actually) in C++ depending on the setting.

[[sound]]

>   

Reviewer: Is this newline necessary?

    de = Halten Sie sich rechts.
    el = Μείνετε δεξιά.
    es = Manténgase a la derecha.

>     es-MX = Manténgase a la derecha.

Reviewer: Will our engine work if duplicate sub-language strings are removed if they are equal to the main language (e.g. es == es-MX, pt == pt-BR, etc.)? Other strings work fine. I would prefer to remove duplicate sub-language translations to keep it simpler.

Author: We'll find out! Certain TTS-specific words like "onto" do need to be specified but I've removed all other duplication.

>     hu = Jobbra tartson.

Reviewer 2: es= Manténgase a la derecha.

Author: It was already this value.

    tr = Sağdan devam edin.
    uk = Тримайтеся правіше.
    vi = Đi về phía phải.
    zh-Hans = 向右行驶
    zh-Hant = 向右走。

  [make_a_slight_right_turn_street]
    comment = Optionally used inside dist_direction_onto_street in position 3 when the normal make_a_slight_right_turn is insufficient.
    nl = de flauwe bocht naar rechts

  [make_a_slight_right_turn_street_verb]

>     comment = Optionally used inside dist_direction_onto_street in position 5 when make_a_slight_right_turn_street needs to be split into two parts.

Author: @1ec5 thanks for the review Minh! We are using format strings in the sense of `Turn $1 onto $2` etc which is about as flexible as we can get without me majorly rearchitecting the entire app's i18n as my first (long delayed) PR.

We're moving towards Weblate as fast as we can but the core issue is that the developer workflow is via these text files which are actually not a supported format. They look like INIs but most platforms don't support all strings (all languages) in one file: they expect one file per language. So the most direct fix is to just abandon the txt file format entirely and change the per-platform string files directly... but that's a big change for developers and core contributors used to doing it the other way for so long.

>     comment = Optionally used inside dist_direction_onto_street in position 5 when make_a_slight_right_turn_street needs to be split into two parts.

Reviewer 3: > We are using format strings in the sense of `Turn $1 onto $2` etc which is about as flexible as we can get without me majorly rearchitecting the entire app's i18n as my first (long delayed) PR.

“Turn $1 onto $2” would be fine, but the current approach seems to be literally “$1 $2 $3 $4”, where $1 is “Turn” and $3 is “onto”. That’s no different than concatenating Lego strings together. Placeholders are for dynamic content, e.g., street names, not static strings. The challenge OSRMTI faced was that we had to choose between treating words like “right” as dynamic content (harder for some languages to integrate into the overall sentence without custom code) or inlining them as static content (resulting in a proliferation of format strings, because these words can’t be reused as easily). The current approach really leans into the former.

If you’re open to addressing this issue as part of this PR, I think it would be a matter of inlining the existing translations into the format string as a one-time operation. Then the issue of null slots goes away, because the translator can simply omit a word in the format string, confident that another language’s translation won’t take its place.

>     comment = Optionally used inside dist_direction_onto_street in position 5 when make_a_slight_right_turn_street needs to be split into two parts.

Author: @1ec5 I hear you, but I think you're underestimating the complexity of the format string -- which does change, as some languages have different subject-verb-adjective-object ordering. We could certainly move some logic out of the locale strings and into the program, but honestly the most stable fix would be to just have some locale strings/files behave differently than others (maybe a marker or argument that disables the english fallback logic etc.)

Merging some of these strings into the formatter to remove the formatter would likely increase duplication not decrease: previous versions of this PR had something like `$1 $2 onto $3` until we realized that "onto" actually needs to itself be dynamic based on a few different factors.

It's true that we can't very easily reuse any part of "take the fifth exit" across all languages due to counting issues, what constitutes an "exit," how to conjugate "take", etc, but that's not the biggest problem the app faces so I'm not super worried. It is a future goal to support arbitrary distances and countings, like "in `13` meters, turn right", which will put all of this to the test especially with that fun fun fun pluralization, and that's being discussed and maybe we do move to something like Fluent later -- cc @pastk 

For now, this PR is about including the street name in the turn instructions, and I fear if we stray past the minimal changeset to accomplish that goal then it'll be 2026 before this feature gets released.

>     comment = Optionally used inside dist_direction_onto_street in position 5 when make_a_slight_right_turn_street needs to be split into two parts.

Reviewer 3: > I hear you, but I think you're underestimating the complexity of the format string -- which does change, as some languages have different subject-verb-adjective-object ordering.

I don’t think I am. “$1 $2 onto $3” can be translated as “Onto $3 by $2ing $1” if necessary. What you’d lose is the ability to tell that “onto” is the preposition that needs to be inclined differently depending on $3’s gender, so you’d have to have a different string per gender or some form of markup within the string itself.

> Merging some of these strings into the formatter to remove the formatter would likely increase duplication not decrease: previous versions of this PR had something like `$1 $2 onto $3` until we realized that "onto" actually needs to itself be dynamic based on a few different factors.
> 
> It's true that we can't very easily reuse any part of "take the fifth exit" across all languages due to counting issues, what constitutes an "exit," how to conjugate "take", etc, but that's not the biggest problem the app faces so I'm not super worried.

Right, my point is that redundant, unreusable strings aren’t necessarily such a problem that you need to avoid them so aggressively. It’s OK to make translators translate more strings if that gives them more control over the final output. OSRMTI struck a balance that translators found somewhat daunting at first, but it isn’t nearly as burdensome as in Valhalla, which does minimal substitution or concatenation. Sometimes OSRMTI contributors would complain about having to translate 300 strings, but once they got started on the first 20%, Transifex’s translation memory kicked in, speeding up the remaining 80%.

Fluent (and a work-in-progress industry standard, [MessageFormat2](https://github.com/unicode-org/message-format-wg/)) would somewhat address this tradeoff, so that translators would work with something streamlined and the redundant strings would only occur in the built product. Translatewiki.net supports an even more powerful templating language as part of its message format, but that’s less useful for mobile platforms where there’s no tooling to interpret that language.

It isn’t for me to insist on reworking the translation units as part of this PR. Going forward, there will be plenty of opportunities to experience the limitations of the current approach and evaluate alternatives (not that the alternatives are unalloyed goods either).

>     comment = Optionally used inside dist_direction_onto_street in position 5 when make_a_slight_right_turn_street needs to be split into two parts.

Author: The instant we get any more complex than this I fully agree; as you can see Hungarian was a whole thing by itself and I'm not sure Fluent helps much with vowel harmony especially for unknown variables (street names).

But like I said there is a desire to support dynamic numbers and that's a prime Fluent use case; we're already in the realm of silliness with 11 strings for 1-11 roundabout exits. I see the value, but I think it's a project for a future day

    tr = Sağdan devam edin.
    uk = Тримайтеся правіше.
    vi = Đi về phía phải.
    zh-Hans = 向右行驶
    zh-Hant = 向右走。

  [make_a_slight_right_turn_street]
    comment = Optionally used inside dist_direction_onto_street in position 3 when the normal make_a_slight_right_turn is insufficient.

>     nl = de flauwe bocht naar rechts

Reviewer: Can you please remind me what happens if there is no en and other translations? Will an empty string be returned for all languages except nl?

And if there is a `en` translation, then it will be returned for any language, right?

Author: Correct, adding an `en` here would require that all languages *not* having a unique string for "make a slight right turn onto $name" (different from english anyway) would need to be specified as blank here.

    zh-Hans = 向右急转。
    zh-Hant = 向右急轉。

>   [make_a_right_turn_street]

Reviewer: Does it make sense to embed this translation into dist_direction_onto_street instead of defining it separately?

Author: No in fact we need to make a bunch of other _street strings, probably one for every possible direction. In at least Japanese and Dutch we have a situation where a command like turn left or take the 3rd exit has different grammar from a preposition like turn left onto Main Street, it'll end up being more like "left onto Main Street make a turn." So we need separate strings for that and only that case.

Author: If I'm really lucky, "make a turn" and "take an exit" won't be additional complications requiring even more assembled particles. Will have to double check with the JA and NL experts. But yeah duplicating every direction for every language seems excessive.

Reviewer: Also, think about easier support and translation to other languages. Sometimes it is easier to write several similar strings than to understand what these special characters mean and how your algorithm combines them together.

Author: Absolutely, if it gets any more complex or if people struggle then I'll do it. I just don't want to give dozens of people dozens of things to duplicate if it can easily be avoided. Placeholders and fragments in i18n are pretty common.

Reviewer: How many almost identical strings will be now if you remove one placeholder?

Author: My proposed solution is that we only duplicate directional strings (with a _street suffix) for those languages who have different syntax for commands versus prepositional phrases, like Japanese and Dutch. 90% of languages will need no such duplication so there will only be a few dozen new strings:

```
  [make_a_slight_right_turn]
    en = Keep right.
  [make_a_slight_right_turn_street]
    en = keep right

  [make_a_right_turn]
    en = Turn right.
  [make_a_right_turn_street]
    en = turn right

  [make_a_sharp_right_turn]
    en = Sharp right.
  [make_a_sharp_right_turn_street]
    en = sharp right

  [enter_the_roundabout]
    en = Enter the roundabout.
  [enter_the_roundabout_street]
    en = enter the roundabout

//...

  [destination]
    en = You’ll arrive.
  [destination_street]
    en = you’ll arrive

  [dist_direction_onto_street]
    en = %1$s %2$s onto %3$s
```

to do your proposed method of removing the %2 directional placeholder, we'd remove the new dist_direction_onto_street translation but need to duplicate every directional string in every language; the sounds.txt file will be 2x larger if not 3x or 4x to handle some future cases like the desired "exit 123 for city center toward main st" alternative to the current "exit onto 123, city center, main street":

```
  [make_a_slight_right_turn]
    en = Keep right.
  [in_dist_make_a_slight_right_turn_street]
    en = %1$s keep right onto %3$s

  [make_a_right_turn]
    en = Turn right.
  [in_dist_make_a_right_turn_street]
    en = %1$s turn right onto %3$s

  [make_a_sharp_right_turn]
    en = Sharp right.
  [in_dist_make_a_sharp_right_turn_street]
    en = %1$s sharp right onto %3$s

  [enter_the_roundabout]
    en = Enter the roundabout.
  [in_dist_enter_the_roundabout_street]
    en = %1$s enter the roundabout onto %3$s

//...

  [destination]
    en = You’ll arrive.
  [in_dist_destination_street]
    en = %1$s you’ll arrive onto %3$s
```

I think the best solution is the one currently coded. It's not that uncommon to have multiple placeholders as long as we communicate with our translators properly, it's not too hard to program for, and it opens up a lot of excellent future language possibility.

Reviewer: Ok, let's polish, cleanup, and test it properly then. Very good comments and examples for these translations are needed in the sounds.txt for other translators too.

    comment = Used inside formatted phrases like dist_direction_onto_street ("in 500 meters, turn right onto Main Street")

>     ja = 右折し

Reviewer: Default en translation should be always specified.

Author: This will be the one exception to that rule, I meant to ask what's the safeguard for missing translations but the implemented logic is that we look for a _street translation and if it doesn't exist we fall back to the plain translation. English has no need for separate grammar inside a prepositional phrase but a few languages do.

    tr = Tam sağa dönün.
    uk = Різко вправо.
    vi = Rẽ ngoặt bên phải.
    zh-Hans = 向右急转
    zh-Hant = 向右急轉。

  [make_a_sharp_right_turn_street]
    comment = Optionally used inside dist_direction_onto_street in position 3 when the normal make_a_sharp_right_turn is insufficient. For most languages that don't need this, the value should be the word NULL

>     en = NULL

Author: @biodranik we just had a big long conversation about this and this was the solution: it *could* be empty, but then your locale file validator would complain, so explicitly setting it to the string `"NULL"` and checking for that is the only sane workaround I can imagine

    tr = Kavşaktan çıkın.
    uk = З'їзд з кільцевої транспортної розв'язки.
    vi = Ra khỏi vòng xoay.
    zh-Hans = 驶出环岛
    zh-Hant = 退出環島。

>   [make_a_slight_left_turn]

Reviewer 2: es = Manténgase a la izquierda.

Author: It was already this value.

    zh-Hans = 向左急转。
    zh-Hant = 向左急轉。

>   [make_a_left_turn_street]

Reviewer: Does it make sense to embed this translation into dist_direction_onto_street instead of defining it separately?

    tr = Düz ilerleyin.
    uk = Їдьте прямо.
    vi = Đi thẳng.
    zh-Hans = 直行
    zh-Hant = 直行。

>   [exit]

Reviewer 2: es = Salga.

    da = Afkørsel.
    de = Nehmen Sie die Ausfahrt.
    el = Εξοδος.

>     es = Salga

Reviewer: Are you sure that no dot is needed here and below?

Author: It doesn't really matter for something like this because it's either always at the end of a sentence anyway or my logic removes terminal dots when it's in the middle of a sentence.

Before with a dot:

`In 500 feet, Exit.`
`Exit.`

Now with or without a dot:

`In 500 feet, Exit onto Main Street`
`Exit onto Main Street`

TTS is almost always smart enough to still sound authoritative when given a sentence fragment without terminal punctuation, so honestly I'd prefer having no periods at all but it really doesn't matter. I've re-added it here.

    pl = na
    pt = para
    pt-BR = para
    ro = pe
    ru = на
    sk = na
    sv = in på
    sw = kwenye
    th = เข้าสู่
    tr = yönünde
    uk = на
    vi = vào
    zh-Hans = 上
    zh-Hant = 上

  [take_exit_number]
    comment = Used in dist_direction_onto_street instead of "Exit" when there's an exit number after. Replaces the "onto" string as well. (There is no take_exit_number_street since this string is only used when announcing street names anyway.)
    en = Take exit
    ar = خروج إلى
    be = З'езд на
    ca = Agafa la sortida
    cs = výstup na
    da = Afkørsel ad
    de = Nehmen Sie die Ausfahrt auf
    el = Εξοδος στην
    es = Tome la salida
    eu = Irteera
    fa = خروجﺐﻫ
    fi = Poistu tielle
    fr = Sortie sur
    hi = बाहर जाएं पर
    hr = Izlaz na
    hu = Hajtson ki a
    id = Keluar ke
    it = Uscita sulla
    ja = 出口を出る
    ko = 출구
    mr = बाहेर पडा वर
    nb = Ta avkjøringen inn på
    nl = Verlaat naar
    pl = Wybierz zjazd
    pt = Siga pela saída
    pt-BR = Pegue a saída
    ro = Ieșire pe
    ru = Съезд на
    sk = Výjazd na
    sv = Utgång in på
    sw = Kuondoka kwenye
    th = ทางออก เข้าสู่
    tr = çıkın yönünde
    uk = З'їзд на

>     vi = Lối thoát vào

Reviewer 3: “Lối thoát” means “an escape”, and “vào” would introduce a street name, not an exit number.

```suggestion
    vi = Đi theo lối ra số
```

Author: @1ec5 all new strings that haven't had a human reviewer in the comments here are machine generated. You can see the list of language teams at the bottom and who's proposed changes; any and all help is definitely appreciated.

Reviewer 3: Translation change request:
```
# data/strings/sound.txt
  [exit]
-     vi = Lối thoát.
+     vi = Đi theo lối ra
  [take_exit_number]
-     vi = Lối thoát vào
+     vi = Đi theo lối ra số
  [dist_direction_onto_street]
-     vi = %1$s %2$s %3$s %4$s
+     vi = undefined
```

    hu = Hajtson ki!
    id = Keluar.
    it = Uscita.
    ja = 出口です
    ko = 출구.
    mr = बाहेर पडा.
    nb = Ta avkjøringen.
    nl = Verlaat.

>     pl = Wybierz zjazd

Author: @Acrobot line 490 now says `    pl = Zjedź zjazdem.` per your upstream changes, when this PR is merged it will only be used for the definitive stand-alone command like "in 100 meters, exit." Line 570 above for [take_exit_number] will be used in the fuller case of "in 100 meters, take exit 123, Main Street, Queen's Expressway, London" as opposed to "in 100 meters, turn right onto Main Street, Queen's Expressway, London"

    hu = Hajtson ki!
    id = Keluar.
    it = Uscita.

>     ja = 出口です

Reviewer: I also see that other translations also missing a dot. I remember that the TTS engine changes pronunciation depending on the dot presence.

Author: It'd be interesting to hear if that was the case, I haven't noticed it in my tests.

    en = .
    ar = .
    be = .
    cs = .
    da = .
    de = .
    el = .
    es = .
    es-MX = .
    eu = .
    fa = .
    fi = .
    fr = .
    hi = .
    hr = .
    hu = .
    id = .
    it = .
    ja = .
    ko = .
    mr = .
    nb = .
    nl = .
    pl = .
    pt = .
    pt-BR = .
    ro = .
    ru = .
    sk = .
    sv = .
    sw = .
    th = .
    tr = .
    uk = .
    vi = .
    zh-Hans = .
    zh-Hant = .

  [take_exit_number]
    comment = Used in dist_direction_onto_street instead of "Exit" when there's an exit number after. Replaces "onto" unless onto_exit_number is specified.
    en = Take exit
    ar = خروجﺈﻟﻯ
    be = З'езд на
    ca = Agafa la sortida
    cs = výstup na
    da = Afkørsel ad
    de = Nehmen Sie die Ausfahrt auf
    el = Εξοδος στην
    es = Tome la salida
    es-MX = Tome la salida
    eu = Irteera
    fa = خروجﺐﻫ
    fi = Poistu tielle
    fr = Sortie sur
    hi = बाहर जाएं पर
    hr = Izlaz na
    hu = Kilépés a-re
    id = Keluar ke
    it = Uscita sulla
    ja = 出口ですに入ります
    ko = 출구
    mr = बाहेर पडा वर
    nb = Ta avkjøringen inn på
    nl = Verlaat naar
    pl = Wybierz zjazd na
    pt = Saia para

>     pt-BR = Pegue a saída para

Reviewer: ?

>     pt-BR = Pegue a saída para

Author: Added, just didn't see it somehow

    tr = çıkın.
    uk = З'їзд.
    vi = Lối thoát.
    zh-Hans = 驶出
    zh-Hant = 出口.

  [onto]
    comment = Used in dist_direction_onto_street in position 3 unless take_exit_number or onto_exit_number are specified.
    en = onto
    ar = إلى
    be = на
    cs = na
    da = ad
    de = auf
    el = στην
    es = hacia
    es-MX = hacia

>     eu =

Reviewer: Are empty strings expected? If yes, then it's better to mention in a comment.

Author: Some languages don't appear to have an "onto" word so yes sometimes null is possible. I will update the comment.

Author: Tested and commented

    fa = ﺐﻫ
    fi = tielle
    fr = sur
    hi = पर
    hr = na
    hu = a-re
    id = ke
    it = sulla
    ja = 
    ko =
    mr = वर
    nb = inn på
    nl = naar
    pl = na
    pt = para
    pt-BR = para
    ro = pe
    ru = на
    sk = na
    sv = in på
    sw = kwenye
    th = เข้าสู่
    tr = yönünde
    uk = на
    vi = vào
    zh-Hans = 上
    zh-Hant = 上

  [onto_exit_number]

>     comment = TODO: currently unused, consider removing. Used in dist_direction_onto_street instead of "onto" when there's an exit number after. Use "." for none

Reviewer: Why is it not removed?

Author: Was holding it open just in case another language reviewer came along needing it; removing.

  [take_exit_number]
    comment = Used in dist_direction_onto_street instead of "Exit" when there's an exit number after. Replaces "onto" unless onto_exit_number is specified.
    en = Take exit
    ar = خروج إلى
    be = З'езд на
    ca = Agafa la sortida
    cs = výstup na
    da = Afkørsel ad
    de = Nehmen Sie die Ausfahrt auf
    el = Εξοδος στην
    es = Tome la salida
    es-MX = Tome la salida
    eu = Irteera
    fa = خروجﺐﻫ
    fi = Poistu tielle
    fr = Sortie sur
    hi = बाहर जाएं पर
    hr = Izlaz na
    hu = Kilépés a-re
    id = Keluar ke
    it = Uscita sulla
    ja = 出口を出る
    ko = 출구
    mr = बाहेर पडा वर
    nb = Ta avkjøringen inn på
    nl = Verlaat naar
    pl = Wybierz zjazd
    pt = Saia para
    pt-BR = Pegue a saída para
    ro = Ieșire pe
    ru = Съезд на
    sk = Výjazd na
    sv = Utgång in på
    sw = Kuondoka kwenye
    th = ทางออก เข้าสู่
    tr = çıkın yönünde
    uk = З'їзд на
    vi = Lối thoát vào
    zh-Hans = 驶出 上
    zh-Hant = 出口 上

  [take_exit_number_street_verb]
    comment = Optionally used inside dist_direction_onto_street in position 5 when take_exit_number needs to be split into two parts. (There is no take_exit_number_street)

>     ja = 

Reviewer: Are empty strings really needed?

Author: I can check what happens when I take them out, I am somewhat concerned that the automatic fallback logic might put i.e. English words in a Japanese translation if English uses a word like "onto" and Japanese doesn't. The current logic is tested and intentional to say we tested a string in this location but want it explicitly blank.

Reviewer: Then it should be explicitly mentioned in the comment above.

Author: Per other comment, I've tested and explicitly commented.

    el = στην
    es = hacia
    es-MX = hacia
    eu =
    fa = ﺐﻫ
    fi = tielle
    fr = sur
    hi = पर
    hr = na
    hu = a
    id = ke
    it = sulla
    ja = 
    ko =
    mr = वर
    nb = inn på
    nl = naar
    pl = na
    pt = para
    pt-BR = para
    ro = pe
    ru = на
    sk = na
    sv = in på
    sw = kwenye
    th = เข้าสู่
    tr = yönünde
    uk = на
    vi = vào
    zh-Hans = 上
    zh-Hant = 上

  [take_exit_number]
    comment = Used in dist_direction_onto_street instead of "Exit" when there's an exit number after. Replaces the "onto" string as well. (There is no take_exit_number_street since this string is only used when announcing street names anyway.)
    en = Take exit
    ar = خروج إلى
    be = З'езд на
    ca = Agafa la sortida
    cs = výstup na
    da = Afkørsel ad
    de = Nehmen Sie die Ausfahrt auf
    el = Εξοδος στην
    es = Tome la salida
    es-MX = Tome la salida
    eu = Irteera
    fa = خروجﺐﻫ
    fi = Poistu tielle
    fr = Sortie sur
    hi = बाहर जाएं पर
    hr = Izlaz na
    hu = Hajtson ki a
    id = Keluar ke
    it = Uscita sulla
    ja = 出口を出る
    ko = 출구
    mr = बाहेर पडा वर
    nb = Ta avkjøringen inn på
    nl = Verlaat naar

>     pl = Wybierz zjazd

Author: @Acrobot @pm4rcin I'm happy to defer to whatever consensus exists, I just don't speak Polish so I have to judge based on your discussion. Check out https://zyphlar.github.io/organicmaps-locale-viewer/?lang=pl for some example sentences, you can click to edit and suggest changes. What seems best to both of you?

>     pl = Wybierz zjazd

Reviewer 4: @zyphlar from my look it all seems good right now. I have checked your webpage and everything looks good now. So I'm patiently waiting for it to be merged because it is taking too long and it would be good starting point to implement more advanced instructions as discussed throughout whole PR. :)

>     pt = Apanhe a saída

Reviewer 5: ```suggestion
    pt = Siga pela saída
```

    tr = çıkın.
    uk = З'їзд.
    vi = Lối thoát.
    zh-Hans = 驶出
    zh-Hant = 出口.

  [onto]

>     comment = Used in dist_direction_onto_street in position 3 unless take_exit_number is specified. For languages that don't have an "onto" word, it must be specified as blank for that language.

Reviewer: ```suggestion
    comment = Used in dist_direction_onto_street in position 3 unless take_exit_number is specified. For languages that don't have an "onto" word (e.g. eu, ja, ko), the value should be empty.
```

    tr = Hedef noktasına vardınız.
    uk = Ви прибули.
    vi = Bạn đã đến.

>     zh-Hans = 您已到达目的地

Reviewer: Dots were removed intentionally, right? If yes (and if it sounds better without a dot in Chinese), then please also remove dots in corresponding zh-Hant translations.

Author: I'll just re-add them. If it makes a difference for some people's TTS then we can keep them, my code handles them properly. This was just not-caring more than intention.

    de = Sie sind angekommen.
    el = Έχετε φτάσει.
    es = Ha llegado a su destino.

>     es-MX = Ha llegado a su destino.

Reviewer: Do es-MX and pt-BR correctly take es/pt translation if it's equal and missing?

Author: I haven't tested this, `es-MX` appeared on my radar during development and I wanted to stub it out / handle it. AFAIK es-MX is currently equal to es, however pt-BR is definitely not equal to pt in all cases. I can remove either/or if you desire, I don't intend a lot of changes there.

Reviewer 5: Please don't change PT and PT-BR, I've translated and revised them. Yes, there are some differences.

Author: @matheusgomesms can you review the PR as it currently looks and tell me if the current changes are correct? https://github.com/organicmaps/organicmaps/pull/3130/files/00cb82086690e58e6edde753ed166cbe5675f537#diff-631491c869fee64c6cc86150edcfb7f52b23d306aee68d05816f8a141727e022

The only additional change I plan to make tonight is the change you requested on September 8:

```diff
   [take_exit_number]
     en = Take exit
     ar = خروج إلى
     be = З'езд на
@@ -609,8 +569,8 @@
     nb = Ta avkjøringen inn på
     nl = Verlaat naar
     pl = Wybierz zjazd
-    pt = Saia para
-    pt-BR = Pegue a saída para
+    pt = Apanhe a saída
+    pt-BR = Pegue a saída
```

Reviewer 5: Great, thanks! I got confused with many comments :P

I checked the Portuguese community, and it seems that the best approach here is (I guess already implemented on lines 572-573):

pt = Apanhe a saída
pt-BR = Pegue a saída

    zh-Hans = 然后
    zh-Hant = 然後

  [dist_direction_onto_street]
    comment = %1$s is the distance like in_500_meters, %2$s is the direction like make_a_right_turn_street or make_a_right_turn if unavailable, and %3$s is the name/number of the street/exit.

>     en = %1$s %2$s onto %3$s

Reviewer: It looks like %2 can be directly embedded here instead of referencing an external string, right?

Author: No, it's replaced by any direction like turn left, turn right, make a slight right turn, take the 3rd exit, etc. We'd have to duplicate this positional format across every permutation and for 90% of languages it's just unnecessary

    zh-Hans = 然后
    zh-Hant = 然後

  [dist_direction_onto_street]
    comment = %1$s is the distance like in_500_meters, %2$s is the direction like make_a_right_turn_street or make_a_right_turn if unavailable, %3$s is "onto" unless the turn is an exit number, %4$s is the name/number of the street/exit, and %5$s is an optional turn verb particle

>     en = %1$s %2$s %3$s %4$s

Reviewer: Does it make sense to copypaste everything? Shouldn't the en translation work by default for all languages?

Author: I don't know, but if en fallback "just works" then it's a big problem for other strings where I do logic like `if (string myTrans = getTranslation("dist_direction_onto_street", locale)){ outputDistDirOntoStr(myTrans, myString); }` -- automatic fallback would say "oh sure, we have your string. You wanted English right?" when I'm explicitly checking for the string in *my* locale. And if en fallback doesn't "just work" then the answer to your question is no. I've programmed and tested this as written, I'd have to go back and check assumptions to see what happens if/when translation strings are removed or when a string for `en` is defined but not the user's language. (It just so happens that many of our edge cases are in unrelated languages like `nl` and `ja` and not "the default" `en` so the case hasn't been tested.)

Reviewer: It's better not to guess, but check/test, and make it explicit in comments/documentation. Where the fallback is ok, copy-paste is not needed.

Author: @biodranik I've confirmed that fallback is active when `en` is specified. For this string I've deduplicated and specified that all languages inherit from `en` by default, and for `[onto]` I've specified that an empty string is required for some languages in order to prevent nonsense. (generate_localizations would create strings like "一キロ先 Broadway Avenue onto 右折です" otherwise, where Japanese requires no "onto" or equivalent in this position, especially not in English)

    fa = در پنجاه قدمی
    fi = Viidenkymmenen jalan päässä
    fr = Dans cinquante pieds

>     hi = पचास फीट में

Reviewer 6: I'm not familiar with Hindi, but is it ok here and everywhere in PR for _hi_ ?

Author: Yes the `।` character is a period in Hindi, so that has been removed and is the only change.

`पचास फीट में। दायीं ओर मुड़ें। पर highway 123` is translated as `In fifty feet. Turn right. on highway 123`

`पचास फीट में दायीं ओर मुड़ें पर highway 123` is translated as `turn right in fifty feet onto highway 123`

>     hu = Kamera elöl

Reviewer 2: es = Cámara próxima

Make it the same as es-MX, it sounds better
